apis/cloudflare/src/attachment-wrapper.ts

```diff
@@ -0,0 +1,165 @@
+import { isArray, isObject } from "@braintrust/core";
+import {
+  AttachmentReference,
+  braintrustAttachmentReferenceSchema,
+} from "@braintrust/core/typespecs";
+import { getBase64Parts, isBase64File } from "@braintrust/local/functions";
+import {
+  Attachment,
+  BraintrustState,
+  deserializePlainStringAsJSON,
+  ReadonlyAttachment,
+} from "braintrust";
+
+// This is a hack that makes it easy on the typesystem to deal with the payload which
+// doesn't change types.
+export function replacePayloadWithAttachments<T>(
+  data: T,
+  state: BraintrustState | undefined,
+  attachments: Record<string, AttachmentReference>,
+): T {
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  return replacePayloadWithAttachmentsInner(data, state, attachments) as T;
+}
+
+function replacePayloadWithAttachmentsInner(
+  data: unknown,
+  state: BraintrustState | undefined,
+  attachments: Record<string, AttachmentReference>,
+): unknown {
+  if (isArray(data)) {
+    return data.map((item) =>
+      replacePayloadWithAttachmentsInner(item, state, attachments),
+    );
+  } else if (isObject(data)) {
+    return Object.fromEntries(
+      Object.entries(data).map(([key, value]) => [
+        key,
+        replacePayloadWithAttachmentsInner(value, state, attachments),
+      ]),
+    );
+  } else if (typeof data === "string") {
+    if (attachments[data]) {
+      return attachments[data];
+    } else if (
+      isBase64File(
+        data,
+        (mimeType) =>
+          mimeType.startsWith("image/") || mimeType === "application/pdf",
+      )
+    ) {
+      const { mimeType, data: arrayBuffer } = getBase64Parts(data);
+      const filename = `file.${mimeType.split("/")[1]}`;
+      return new Attachment({
+        data: arrayBuffer,
+        contentType: mimeType,
+        filename,
+        state,
+      });
+    } else {
+      return data;
+    }
+  } else {
+    return data;
+  }
+}
+
+export function replaceAttachmentsInPayload<T>(
+  data: T,
+  state: BraintrustState | undefined,
+  attachments: Record<string, AttachmentReference>,
+): Promise<T> {
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  return replaceAttachmentsInPayloadInner(
```

Comment:
fyi, you can just do `Promise.resolve(replaceAttachmentsInPayloadInner(...))`

Reply:
wouldn't map to the generic type `T` though. we can iterate on the wrapper logic in a followup